### DISCUSSION TITLE
SagemakerProcessingOperator ThrottlingException

### DISCUSSION THREAD

**Discussion by jimmycfa** [apache/airflow#18375] [Q&A]
**Apache Airflow version**: 2.0.2

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): NA

**Environment**: MWAA and Locally

- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): NA
- **Kernel** (e.g. `uname -a`): NA
- **Install tools**: NA
- **Others**: NA

**What happened**:

When calling the SagemakerProcessingOperator sometimes get: "botocore.exceptions.ClientError: An error occurred (ThrottlingException)" due to excessive ListProcessingJobs operations.

**What you expected to happen**:

The job should have started without timing out. I believe one fix would be to use the `NameContains` functionality of boto3 [list_processing_jobs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.list_processing_jobs) so you don't have to paginate as is occurring [here](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/sagemaker.py#L916). 

**How to reproduce it**:

If you incrementally create Sagemaker Processing jobs you will eventually see the Throttling as the pagination increases.

**Anything else we need to know**:

This looks like it is happening when the account already has a lot of former Sagemaker Processing jobs.

**Comment by jimmycfa** [comment]
Realized this is already happening. Sorry! We are still seeing this issue but going to close until I do some more investigating.

**Comment by jimmycfa** [comment]
For posterity sake we are using boto3==1.17.99. This actually appears to be an issue with the way that NameContains filter gets applied:

The NameContains is getting passed into list_processing_jobs but it doesn't actually filter on the entire set of ProcessingJobs. It appears to filter per batch of 100 so you still end up calling the list_processing_jobs in that SagemakerOperator 30+ times back to back. Another way of saying this is if I specify NameContains in the list_processing_jobs with a job name that doesn't exist and I have over 3500 processing jobs it will return an empty set of ProcessingJobSummaries BUT still includes a NextToken. It will do this 35 more times as the max results = 100 for that call and you likely run into Throttling issues.

I believe the expected behavior of that boto3 call should be the NameContains filter should be being applied to the entire set of jobs and then returning results vs per batch so that the first call through returns an empty set for ProcessingJobSummaries and NO NextToken.

I'm going to reopen but this does appear to be a boto3 issue.

Our current workaround was to update the `aws_default` connection in Admin->Connections and add the following to Extr:
```json
{
   "config_kwargs":{
      "retries":{
         "max_attempts":10,
         "mode":"standard"
      }
   }
}
```

**Comment by potiuk** [comment]
Since we have a solution here - which appears to be applicable by everyone, I think it should be closed. Alternatively maybe you could add the solution above as a comment to the documentation of the SageMaker Operator @jimmycfa ?

**Comment by ChanglinZhou** [comment]
I believe the root cause is sagemaker operator calls boto3 list api to get all the processing jobs, which is a linear search, we should use describe_processing_job instead.

**Reply by jimmycfa** [reply]
This does seem like a better approach: [boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.describe_processing_job) You could use this and put an exception catch for `SageMaker.Client.exceptions.ResourceNotFound`

**Comment by mycaule** [comment]
In Airflow 2.2.2 which is the latest version for AWS MWAA (managed service),
I found out you can use the `check_if_job_exists=False` parameter to by pass the calls.
See my investigation in https://github.com/apache/airflow/discussions/26219

**Comment by Taragolis** [comment]
`boto3` itself supports different exponential backoff retry strategy in case of different Throttling exceptions.

Hovewer by default `boto3` use "super-optimistic" configuration.

How To Change retry strategy:
https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#avoid-throttling-exceptions

**Reply by jimmycfa** [reply]
We tried various approaches like this for the original issue. The issue is more how boto3 does its search, e.g. it has to literally read through every pagination batch to look for an existing job vs doing a search in the backend and doing a single return. No matter how you set the throttling strategy as the number of jobs increases you will hit a throttling exception eventually.

I think the recommended approach if you have a large number of existing Sagemaker jobs in your account would be to use the `check_if_job_exists=False` flag or to update the operator to use the `describe_proccesing_job` function instead.

**Reply by mycaule** [reply]
The config tweak in the Airflow connections didn't work for me (Airflow 2.2.2 latest on AWS MWAA) indeed.

Funny that AWS Operators don't even work out of the box for their own services, and that `check_if_job_exists` param is not documented, I had to go through the source code to learn about it.

You will get an exception anyway, later in the code, in the case you use the same job name twice.

**Reply by Taragolis** [reply]
If `10` retries would not enough that actually operator spam API. Backoff exponential retries it is the general recommendation by Amazon. 
However if it not enough and there is solution exists, that might be better commit it to amazon-provider?

Also it could be another problem for users who use MWAA. Guys from AWS OSS Team approve that MWAA right now do not able to upgrade amazon-provider: https://github.com/apache/airflow/issues/25996#issuecomment-1246834320

> and that check_if_job_exists param is not documented

Does this parameter even exists for `SageMakerProcessingOperator`? Latest [5.1.0 version](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/sagemaker/index.html#airflow.providers.amazon.aws.operators.sagemaker.SageMakerProcessingOperator) and [2.4.0](https://airflow.apache.org/docs/apache-airflow-providers-amazon/2.4.0/_api/airflow/providers/amazon/aws/operators/sagemaker_processing/index.html#airflow.providers.amazon.aws.operators.sagemaker_processing.SageMakerProcessingOperator)

**Reply by mycaule** [reply]
In 2.2.2 yes,
https://github.com/apache/airflow/blob/2.2.2/airflow/providers/amazon/aws/operators/sagemaker_training.py#L99-L101

and 2.4.0,
https://airflow.apache.org/docs/apache-airflow-providers-amazon/2.4.0/_api/airflow/providers/amazon/aws/operators/sagemaker_training/index.html#module-airflow.providers.amazon.aws.operators.sagemaker_training

I didn't used `SageMakerProcessingOperator` so I can't tell. I didn't need it though for `ModelOperator` and `TransformOperator`.

**Reply by Taragolis** [reply]
> In 2.2.2 yes,
https://github.com/apache/airflow/blob/2.2.2/airflow/providers/amazon/aws/operators/sagemaker_training.py#L99-L101

This is reference to Airflow 2.2.2. Airflow package itself do not contain provider packages and version of provider packages, as well as release cycle, independent of Airflow itself (except min airflow version supported)

> I didn't used SageMakerProcessingOperator so I can't tell.

My bad. I assume you just talk about 'SagemakerProcessingOperator ThrottlingException'.

But anyway docstring for `airflow.providers.amazon.aws.operators.sagemaker_training.SageMakerTrainingOperator` contain information about this argument

```
check_if_job_exists  -- If set to true, then the operator will check whether a training job already exists for the name in the config.
```

**Reply by mycaule** [reply]
Should be set to true by default to avoid that automatic bug then.

**Reply by Taragolis** [reply]
> Should be set to true by default to avoid that automatic bug then.

You mean False? Because it set as True by default.

https://github.com/apache/airflow/blob/983993c44dbe2f7768b33c7d1606b74aaeddca52/airflow/providers/amazon/aws/operators/sagemaker.py#L664-L666

And do you think that ThrottlingExceptions it is a bug?
Or the reason of miss configure something?
Or not proper description in docstring? 

And main thing is this persists in the latest version of the provider?

**Reply by mycaule** [reply]
Yes I mean `False`, 

Yes it is a bug, it is easy to have an account with already hundreds or thousands of completed training jobs. You will always get that bug when using the `SageMakerTrainingOperator` in that case when entering the `list_processing_job` function.

It would need `check_if_job_exists=False` as default value, or fix the code using `describe_training_job` or `describe_processing_job` instead of `list_..._jobs` as said above.
